File: Study Plan/LeetCode 75/0605 Can Place Flowers/CanPlaceFlowers.java
/*  
    Problem
    --------------------------------------------------------------------------------
    605. Can Place Flowers
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    https://leetcode.com/problems/can-place-flowers/solutions/4529149/java-time-complexity-o-n-space-complexity-o-1-beats-98-in-runtime/
    
    - Runtime 1 ms (Beats 98.39% of users with Java)
    - Memory 45.10 MB (Beats 29.62% of users with Java)
*/

class Solution {
    public boolean canPlaceFlowers(int[] flowerbed, int n) {
        for (int plot = 0; plot < flowerbed.length; plot++) {
            boolean isPreviousPlotEmpty = getElementAtOrDefault(flowerbed, plot-1) == 0;
            boolean isCurrentPlotEmpty = flowerbed[plot] == 0;
            boolean isNextPlotEmpty = getElementAtOrDefault(flowerbed, plot+1) == 0;
            // Check if the previous, current and next plots are empty
            if (isPreviousPlotEmpty && isCurrentPlotEmpty && isNextPlotEmpty) {
                // Plant a flower in the plot
                flowerbed[plot] = 1;
                // Decrease remaining flowers
                n--;
                // Early return
                if (n <= 0) return true;
            }
        }
        return n <= 0;
    }
    private static int getElementAtOrDefault(int[] array, int index) {
        if (index >= 0 && index < array.length) {
            return array[index];
        } else {
            return 0;
        }
    }
}

File: Study Plan/LeetCode 75/0605 Can Place Flowers/CanPlaceFlowers.cs
/*  
    Problem
    --------------------------------------------------------------------------------
    605. Can Place Flowers
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 107 ms (Beats 41.56% of users with C#)
    - Memory 48.43 MB (Beats 5.72% of users with C#)
*/

public class Solution
{
    public bool CanPlaceFlowers(int[] flowerbed, int n) 
    {
        for (int plot = 0; plot < flowerbed.Length; plot++)
        {
            // https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.elementatordefault
            
            var isPreviousPlotEmpty = flowerbed.ElementAtOrDefault(plot-1) == 0;
            var isCurrentPlotEmpty = flowerbed[plot] == 0;
            var isNextPlotEmpty = flowerbed.ElementAtOrDefault(plot+1) == 0;

            // Check if the previous, current and next plots are empty
            if (isPreviousPlotEmpty && isCurrentPlotEmpty && isNextPlotEmpty)
            {
                // Plant a flower in the plot
                flowerbed[plot] = 1;
                // Decrease remaining flowers
                n--;
                // Early return
                if (n <= 0) return true;
            }
        }
        
        return n <= 0;
    }
}